### PR TITLE
design: add access log (ALS) design doc

### PR DESCRIPTION
**Pull request by abhide**
Updates: #1691

Signed-off-by: Amey Bhide <amey15@gmail.com>

### CODE REVIEW

File: design/als-design.md
Comment (sunjayBhatia) [inline design/als-design.md:56]: I don't think we currently are able to pass a config file to `contour bootstrap` so everything is currently done with CLI flags and arguments.

File: design/als-design.md
Comment (sunjayBhatia) [inline design/als-design.md:38]: This annotation pattern is an interesting one and not one we have used already in Contour

It has the downside of making `boostrap` and `serve` look through all `ExtensionService` resources on startup, but makes it so we don't have to pass the same config in two places since we need to tell both `serve` and `bootstrap` about the ALS service.

File: design/als-design.md
Comment (sunjayBhatia) [inline design/als-design.md:63]: Some discussion about what happens when someone isn't using `contour bootstrap` to set up their Envoy bootstrap config would be useful. Also some discussion about what happens with the case where the bootstrap does not have the ALS cluster, since the Contour controller sort of has to assume that it is set up correctly.

Comment (sunjayBhatia) [inline design/als-design.md:63]: We could try to push for Access Logging clusters not having to be static, though from previous Envoy issues and discussion it seems the maintainers are not super into that (which makes sense)

**Review by sunjayBhatia — COMMENTED** [review]
Some initial comments

File: design/als-design.md
Comment (skriss) [inline design/als-design.md:25]: It looks like Envoy requires the ALS cluster to be static, not a CDS cluster, is that right? If so, a brief explanation of that as a constraint (and some links) would be helpful since that guides the design.

Comment (sunjayBhatia) [reply design/als-design.md:25]: +1

File: design/als-design.md
Comment (jpeach) [reply design/als-design.md:38]: Yeh, the idea of using an annotation is OK in theory, but in practice I suggest that the name of the annotation should be much more specific (ie. it obviously means "use this as a default log sink"), and it somehow needs to be tied to a specific envoy deployment (there can be more than one in the cluster).

Perhaps this use case can be configured more directly in the Contour flags, or in the operator?

File: design/als-design.md
Comment (stevesloka) [inline design/als-design.md:35]: To do this, the bootstrap command will need a k8s client to first query the api for the ExtensionService. 

To me, setting args (or a config file) is much more simpler for this use-case and the extra indirection of adding an ExtensionService doesn't help but make it more confusing.

**Review by stevesloka — CHANGES_REQUESTED** [review]
Just a small comment on not utilizing an ExtensionService for this implementation. Curious about what others also think.

File: design/als-design.md
Comment (youngnick) [reply design/als-design.md:35]: The good thing about the ExtensionService is that it allows users to tune the timeouts, which will almost certainly be useful if you are using a sidecar.

I think it will also end up really confusing if we end up with a bunch of different ways to configure extra services in Envoy. We are using ExtensionService for auth and rate limiting already, and will probably need to for tracing. Although it's an extra step, it feels to me like it fits better to have the ExtensionService in there.

File: design/als-design.md
Comment (youngnick) [reply design/als-design.md:38]: The point about multiple Envoy deployments is a very good one, perhaps the annotation could be something like:
`projectcontour.io/envoy-log-sink: <string>" where string is a value that's defaulted, but can be tuned individually on the `contour serve` and `contour bootsrap` commands?

The whole problem is that the bootstrap and the contour deployment need to have the same info configured somehow, so we either need to make a way for them to coordinate, or configure things twice, which seems very prone to failure.

File: design/als-design.md
Comment (abhide) [reply design/als-design.md]: fixed.

File: design/als-design.md
Comment (abhide) [reply design/als-design.md]: https://github.com/envoyproxy/envoy/issues/3660

**Review by abhide — COMMENTED** [review]
Thank you for the feedback and review comments. 
I have incorporated the feedback on the review and from the community meeting and addressed the review comments.

File: design/als-design.md
Comment (youngnick) [inline design/als-design.md:41]: Doing ExtensionService like this does mean that there can only ever be one ALS per cluster, and we will need to build a totally different mechanism to do more than one, which seems not ideal.

On the other hand, I suspect that most people doing this are probably going to want to send to a single central log sink anyway, or to localhost. @bgagnon is currently using ALS through some clever tricks, maybe he can check out this proposal.

**Review by youngnick — CHANGES_REQUESTED** [review]
I'm a bit worried that we're painting ourselves into a corner with this one. I'd like to summon @bgagnon to check it out and see what he thinks, as he's actually using ALS already.

File: design/als-design.md
Comment (stevesloka) [reply design/als-design.md]: What if we ditch the `contour bootstrap` initContainer and have the Leader in Contour manage this configmap? This would mean that Contour needs to write a configmap RBAC permission in a namespace, but would allow for a single config + setup.

**Comment by bgagnon** [comment]
> I'm a bit worried that we're painting ourselves into a corner with this one. I'd like to summon @bgagnon to check it out and see what he thinks, as he's actually using ALS already.

Unfortunately, I have since changed jobs and no longer using Contour or Envoy in my day job. That being said, I remain a fan, and can still have a look at the design proposal!

**Comment by abhide** [comment]
> > I'm a bit worried that we're painting ourselves into a corner with this one. I'd like to summon @bgagnon to check it out and see what he thinks, as he's actually using ALS already.
> 
> Unfortunately, I have since changed jobs and no longer using Contour or Envoy in my day job. That being said, I remain a fan, and can still have a look at the design proposal!

Thanks @bgagnon. Looking forward to your feedback

File: design/als-design.md
Comment (skriss) [reply design/als-design.md]: 👍  to leaving the door open for multiple ALS in the future.

**Comment by stevesloka** [comment]
I think your plan with the annotations would work @abhide, but just curious if there's a way to make the setup easier for now until we're sure we want multiple ALS instances to be configured. 

What if we setup an ALS cluster by default in the bootstrap config? Then the static cluster defined in the bootstrap config chooses an EDS cluster to fill the endpoints of the static cluster. 

Then in Contour, you'd configure which ExtensionService (or just Contour config) which maps to the ALS service and Contour would fill in the endpoints of the static cluster already defined in the bootstrap config. Contour would translate the endpoints of the configured cluster in Kubernetes to the static EDS endpoint. 

I'm not sure what happens in Envoy if it tries to send traffic to an ALS cluster with no endpoints (perf issue?). 

This also might need to follow @youngnick's suggestion of this being the "default" for Contour such that it allows more instances to be added later on.

**Comment by youngnick** [comment]
I understand the suggestion here @stevesloka, but having an ALS static cluster created by default in every Contour installation feels like we're working hard to avoid something that we really should do anyway (and that's have a way to configure the bootstrap that's not just command line flags). This feels like a good forcing function for us to do something to fix that up.

The sort of thing we're talking about using an annotation for here is exactly what they're designed for, to allow you to find a specific set of things without knowing their name beforehand.

However, both of the options in this proposal as it stands have a problem that I hadn't raised enough before. That is that both of these require the Envoy pod to have direct access to the Kubernetes apiserver, because the bootstrap will need access, it runs in the Envoy pod, and there is no way to different security contexts within a Pod. This is a significant expansion in the Envoy Pod's security footprint, that does worry me more now I've thought about it.

Our options there are either that we accept that security expansion and implement one of these, or move to a convention-based approach like @stevesloka suggested. I don't like having that we'll be doing such a different flow to anything else, and I'm not sure how we would extend this to allow multiple ALS sinks, but both of those seem like less of a problem than the security footprint increase.

**Comment by abhide** [comment]
@stevesloka 
I agree with @youngnick's comment; that always configuring` a static ALS cluster by default doesn't seem like the right thing to do in this case.
Yes, we would need access to Kubernetes APIServer from the envoy pod as part of bootstrap. As far as RBAC is concerned, allowing access to ExtensionService objects might be a one-level better as compared to granting access to ConfigMaps. 
Reason being ExtesnionService is a Contour CRD whereas Configmap is a more general Kubernetes object that would be used by other services as well. 
Also, ExtensionService has a Status field, that can be used to communicate whether ALS was configured successfully as part of bootstrap and whether it can be used by `contour serve` to configure ALS entries for HTTP filters.

File: design/als-design.md
Comment (abhide) [reply design/als-design.md]: Yes, I would suggest we add 2 annotations that would be specific to ALS
`projectcontour.io/als-admitted: true` -> Indicates whether to configure this ALS
`projectcontour.io/envoy-log-sink: <name>` -> ALS name or default.

**Comment by abhide** [comment]
@youngnick @stevesloka @skriss Seems like we are in consensus about usage of ExtensionService and exposing the ability to configure multiple ALS sinks (for future extensibility)? 
If yes, would like to know the next steps to get  a :+1: on this design so that implementation work could begin.

**Comment by stevesloka** [comment]
I'm always a fan for simpler than extra setup/config. If we're not sure folks would even want multiple ALS instances, then to me, a simpler approach is to have one setting and see what the community says about their needs, then change it and make it more complex later on. Then don't need to deal with the init requiring access to the API server, etc, annotations, etc. 

I think your approach would work @abhide, thanks for putting this together, but I'll leave this up to what other folks think since it seems my initial implementation ideas are different. 

@skriss @sunjayBhatia do you have some time to look over this design doc + comments today?

**Review by sunjayBhatia — COMMENTED** [review]
Some issues that need some more detail written down and worked out in the design I think would be around some error/misuse flows etc. and what happens there. Especially calling out what happens when someone changes something on the ExtensionService, wants to switch between ALS servers, restarts the Envoy daemonset, isn't using `contour bootstrap` to manage their Envoy bootstrap and makes a mistake specifying the ALS cluster, deletes an ALS ExtensionService etc. With all of our other ExtensionService uses we do not have the need to configure a static cluster in the bootstrap config so most of these issues are completely handled within the Contour controller only. However since we now have multiple components interacting, its a little more complicated.

It would also be good to summarize some of the questions and answers/proposed alternatives in discussion threads in this PR in the design as an "Alternatives Considered"

**Review by sunjayBhatia — COMMENTED** [review]
From discussion in office hours today, there is at least one more option to add to the possible implementations:

As we seem to be moving forward with a plan to enable Contour to manage the Envoy data-plane (see https://docs.google.com/document/d/1kOrGqHXULYO6YLxGCOtVyjz_7mfcW7Yi1WXCpQz6_Qo/edit#heading=h.awh08ldngzyk) it may be that we have a future where Contour itself writes out the bootstrap config, which would really simplify things. However, if Contour is not managing the Envoy bootstrap/deployment, we still have an issue.

**Comment by skriss** [comment]
A few comments:

- I don't think we *need* to support multiple ALS sinks initially, I just didn't want to make design choices that would prevent us from doing that down the road if we find out users want it. I'm fine with supporting at most one ALS per Contour/Envoy initially.
- We talked some at office hours today about the possibility of rearchitecting how Contour generates the Envoy bootstrap config, so that instead of generating it from an initContainer in the Envoy daemonset, we have the main `contour serve` command generate it as a ConfigMap when Contour starts up, and then all the Envoys would mount the ConfigMap as a volume. The benefits of this are that (a) the `envoy` service account doesn't need Kubernetes API access; and (b) the ALS config could now go into the Contour config file, so we wouldn't have to use the new annotation pattern to identify the ALS. This approach seems to solve both of these issues nicely, but it is a fairly large change to the startup sequence for Contour+Envoy, and we'd have to think about how to support users who want to bring their own bootstrap, etc.
- Another alternative idea to consider (sorry if we already discussed this): but couldn't the `bootstrap` init container mount the **Contour** config file as a volume, and read ALS config info from there? That would eliminate the need for the annotation; we could just identify the ALS ExtensionService in the config file, similarly to how the rate limit service is configured today, and then both `contour bootstrap` and `contour serve` would read it from there. This shouldn't require the `envoy` service account to have any new RBAC permissions.

**Comment by youngnick** [comment]
> A few comments:
> 
> * I don't think we _need_ to support multiple ALS sinks initially, I just didn't want to make design choices that would prevent us from doing that down the road if we find out users want it. I'm fine with supporting at most one ALS per Contour/Envoy initially.

That was my feeling as well. It felt to me like the "add a static cluster by default" would be hard to move away from, this feels less hard to change our mind.

> * We talked some at office hours today about the possibility of rearchitecting how Contour generates the Envoy bootstrap config, so that instead of generating it from an initContainer in the Envoy daemonset, we have the main `contour serve` command generate it as a ConfigMap when Contour starts up, and then all the Envoys would mount the ConfigMap as a volume. The benefits of this are that (a) the `envoy` service account doesn't need Kubernetes API access; and (b) the ALS config could now go into the Contour config file, so we wouldn't have to use the new annotation pattern to identify the ALS. This approach seems to solve both of these issues nicely, but it is a fairly large change to the startup sequence for Contour+Envoy, and we'd have to think about how to support users who want to bring their own bootstrap, etc.

This idea seems neat, and does away with some of these problems, while also allowing us to do more clever bootstrap things in the future. Supporting users who want to do their own bootstrap seems straightforward, we can have a boolean to turn off managing the configmap, and/or have users change their Envoy deployment to not use the generated/managed ConfigMap.
Overall, a neat idea, that plays well in the case that Contour ends up managing the Envoys as well. +1 from me.

> * Another alternative idea to consider (sorry if we already discussed this): but couldn't the `bootstrap` init container mount the **Contour** config file as a volume, and read ALS config info from there? That would eliminate the need for the annotation; we could just identify the ALS ExtensionService in the config file, similarly to how the rate limit service is configured today, and then both `contour bootstrap` and `contour serve` would read it from there. This shouldn't require the `envoy` service account to have any new RBAC permissions since it's a volume mount, not an API read (this seems to be true from my testing but we should definitely confirm). I guess `envoy` would still need to be able to read `ExtensionServices`, though, unless we also did away with using an ExtensionService at all.

Yes, @abhide and I talked about this one, and the change in the size of the Envoy RBAC still seems like the sticking point for me. Right now, Envoy has *no* Kubernetes RBAC, or even the token mounted in, so there's zero privilege escalation risk. I'd prefer to keep this if we can.

**Comment by youngnick** [comment]
I should add, I think that Sunjay's thoughts about adding more information about the flows for various error cases is very useful, as is some more background in the "Alternatives considered" area.

**Comment by youngnick** [comment]
@abhide, what's your status here? No problem if you're intending to come back to it, but if not, one of us can probably finish this up for you.

**Comment by abhide** [comment]
> @abhide, what's your status here? No problem if you're intending to come back to it, but if not, one of us can probably finish this up for you.

Hey @youngnick I will fix the review comments and push another PR.
Sorry, had been busy lately. Will look into this in a next few days.

**Comment by youngnick** [comment]
Great, thanks @abhide!

**Pull request closed without merging**
